lib/SILOptimizer/Utils/ConstantFolding.cpp

```diff
@@ -1781,6 +1796,17 @@ ConstantFolder::processWorkList() {
       continue;
     }
 
+    if (isApplyOfBuiltin(*I, BuiltinValueKind::IsConcrete)) {
+      if (SILValue v = constantFoldIsConcrete(cast<BuiltinInst>(I))) {
+        // Here, the bulitin instruction got folded, so clean it up.
+        recursivelyDeleteTriviallyDeadInstructions(
+            I, /*force*/ true,
+            [&](SILInstruction *DeadI) { WorkList.remove(DeadI); });
+        InvalidateInstructions = true;
+      }
+      continue;
+    }
+
```

Comment:
Optimizing this builtin works by:
1. Inserting it in the worklist during initializeWorklist, which you've done
2. Adding a call to constantFoldIsConcrete to contentFoldBuiltin, which you've done-- that will be called for everything in the worklist.

So, this extra optimization can just be removed. If I'm right, your tests should still pass.

Reply:
So I ran my tests against this bit being removed and it starts to fail. So apparently this is a necessary component. 🤷‍♂️

Reply:
@atrick gentle ping

Reply:
I dunno, but it definitely doesn't make sense to do the same optimization in two places. To debug it take a minimal .sil test case and run `sil-opt -diagnostic-constant-propagation t.sil`. My reading of the code is that your change to `initalizeWorklist` would push the builtin apply instruction on the worklist, and `constantFoldBuiltin` should get a chance to run on each worklist item.